expression/builtin_grouping.go

```diff
@@ -0,0 +1,186 @@
+	version     uint32
+	groupingIDs map[int64]struct{}
+}
+
+func (b *builtinGroupingSig) SetMetaVersion(version uint32) {
+	b.version = version
+}
+
+func (b *builtinGroupingSig) SetMetaGroupingIDs(groupingIDs map[int64]struct{}) {
+	b.groupingIDs = groupingIDs
+}
+
+func (b *builtinGroupingSig) getMetaVersion() uint32 {
+	return b.version
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{
+		// TODO
+	}
+	return args
+}
+
+func (b *builtinGroupingSig) Clone() builtinFunc {
+	newSig := &builtinGroupingSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.version = b.version
+	newSig.groupingIDs = b.groupingIDs
+	return newSig
+}
+
+func (b *builtinGroupingSig) getMetaGroupingIDs() map[int64]struct{} {
+	return b.groupingIDs
+}
+
+func (b *builtinGroupingSig) getMetaGroupingID() int64 {
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingIDs()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingSig) checkMetadata() error {
+	version := b.getMetaVersion()
+	grouping_ids := b.getMetaGroupingIDs()
+	if version < 1 || version > 3 {
+		return errors.Errorf("Version of meta data in grouping function is invalid. input version: %d", version)
+	} else if (version == 1 || version == 2) && len(grouping_ids) != 1 {
+		return errors.Errorf("Invalid number of groupingID. version: %d, number of groupingID: %d", version, len(b.groupingIDs))
+	}
+	return nil
+}
+
+func (b *builtinGroupingSig) groupingImplV1(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID&metaGroupingID > 0 {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingSig) groupingImplV2(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID > metaGroupingID {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingSig) groupingImplV3(groupingID int64) int64 {
+	grouping_ids := b.getMetaGroupingIDs()
+	_, ok := grouping_ids[groupingID]
+	if ok {
+		return 0
+	}
+	return 1
+}
+
+func (b *builtinGroupingSig) grouping(groupingID int64) int64 {
+	switch b.version {
+	case 1:
+		return b.groupingImplV1(groupingID, b.getMetaGroupingID())
+	case 2:
+		return b.groupingImplV2(groupingID, b.getMetaGroupingID())
+	case 3:
+		return b.groupingImplV3(groupingID)
+	}
+	return 0
+}
+
+// evalInt evals a builtinGroupingSig.
+func (b *builtinGroupingSig) evalInt(row chunk.Row) (int64, bool, error) {
+	err := b.checkMetadata()
+	if err != nil {
+		return 0, false, err
+	}
+
+	groupingID, isNull, err := b.args[0].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+
+	return b.grouping(groupingID), false, nil
+}
+
+func (b *builtinGroupingSig) groupingVec(groupingIds *chunk.Column, rowNum int, result *chunk.Column) {
+	result.ResizeInt64(rowNum, false)
+	resContainer := result.Int64s()
+	for i := 0; i < rowNum; i++ {
+		resContainer[i] = b.grouping(groupingIds.GetInt64(i))
```

Comment:
use swich inside loop seems not very efficient?

Reply:
> use swich inside loop seems not very efficient?

okk, I will fix it.